docs/chymyst01.md

```diff
@@ -386,43 +395,164 @@ Sometimes it appears that repeating input molecules is the most natural way of e
+We expect to receive many molecules `data(x)` with integer values `x`,
+and we need to compute and print the final sum value when no more `data(x)` molecules are present.
+
+Here is an (incorrect) attempt to write chemical laws for this program:
+
+```scala
+val data = m[Int]
+val sum = m[Int]
+join (
+  run { case data(x) + sum(y) => sum(x+y) }, // We really want the first reaction to be high priority
+   
+  run { case sum(x) => println(s"sum = $x") }  // and run the second one only after all `data` molecules are gone.
+)
+data(5) + data(10) + data(150)
+sum(0) // expect "sum = 165"
+
+```
+
+Our intention was to run only the first reaction and to ignore the second reaction as long as `data` molecules are available in the soup.
+The chemical machine does not actually allow us to assign a higher priority to the first reaction.
+But, if we were able to do that, what would be the result?
+
+In a real-life situation, the `data` molecules are going to be injected concurrently by different processes.
+(There wouldn't be much point in making the `data` molecules concurrent if they were all guaranteed to be present at the start of our program: we would have just used an array instead.)
+
+Since these other processes are concurrent and inject `data` molecules at unpredictable times,
+it could happen that the `data` molecules are injected somewhat more slowly than we are consuming them.
+If that happens, there will be a brief interval of time when no `data` molecules are in the soup (although other processes are about to inject some more of them).
+The chemical machine will then run the second reaction, consume the `sum` molecule and print the result, signalling (incorrectly) that the computation is finished.
+Perhaps this failure will _rarely_ happen, -- it unlikely to show up in your unit tests, but at some point it is definitely going to happen in production.
+
+This kind of nondeterminism is the prime reason concurrency is widely regarded as a hard programming problem.
+
+`JoinRun` will actually refuse our attempted program and print an error message before running anything, immediately after we define the chemical laws:
+
+```scala
+val data = m[Int]
+val sum = m[Int]
+join (
+  run { case data(x) + sum(y) => sum(x+y) },
+  run { case sum(x) => println(s"sum = $x") }
+)
+```
+
+```
+Exception: In Join{data + sum => ...; sum => ...}: Unavoidable nondeterminism: reaction data + sum => ... is shadowed by sum => ...
+
+```
+
+The error message means that the reaction `sum => ...` will sometimes prevent `data + sum => ...` from running,
+and the programmer will have no control over this nondeterminism.
+
+The correct way of implementing this problem is to keep track of how many `data` molecules we already consumed,
+and to emit `done` when we reach the total expected number of the `data` molecules.
+The main change is that the `sum` molecule will now carry a tuple of two integers instead of a single integer.
```

Comment:
and most astute FP-inclined readers (such as Joe Barnes, @joescii), unlike Akka, we don't need state to track remaining.